File: samples/java/Bingo/Server/Api/src/main/java/systems/zlink/samples/bingo/server/api/handlers/GetPlayerRecordHandler.java
package systems.zlink.samples.bingo.server.api.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.bingo.server.api.BingoPlayerRecordStore;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.ApiChannel)
public final class GetPlayerRecordHandler
        implements ZLinkRequestHandler<Messages.GetPlayerRecordReq, Messages.GetPlayerRecordRes> {
    private final BingoPlayerRecordStore records;

    public GetPlayerRecordHandler(BingoPlayerRecordStore records) {
        this.records = records;
    }

    @Override
    public CompletionStage<Messages.GetPlayerRecordRes> handle(
            Messages.GetPlayerRecordReq request, ZLinkMessageContext context) {
        var record = records.get(request.getActorId());
        return CompletableFuture.completedFuture(
                BingoMessages.getPlayerRecordRes(record.actorId(), record.wins(), record.losses()));
    }
}

File: samples/kotlin/Bingo/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/api/handlers/GetPlayerRecordHandler.kt
package systems.zlink.samples.kotlin.bingo.server.api.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.samples.kotlin.bingo.server.api.BingoPlayerRecordStore
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.shared.contracts.GetPlayerRecordReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.GetPlayerRecordRes

@ZLinkHandlerGroup(SampleNames.ApiChannel)
class GetPlayerRecordHandler(private val records: BingoPlayerRecordStore) :
    ZLinkSuspendingRequestHandler<GetPlayerRecordReq, GetPlayerRecordRes> {
    override suspend fun handle(
        request: GetPlayerRecordReq,
        context: ZLinkMessageContext,
    ): GetPlayerRecordRes {
        val record = records.get(request.actorId)
        return GetPlayerRecordRes(record.actorId, record.wins, record.losses)
    }
}
